### ISSUE TITLE
REINFORCE Correction

### ISSUE THREAD

**Issue by IbrahimSobh** [udacity/deep-reinforcement-learning#42]
Hello, 

In deep-reinforcement-learning/reinforce/REINFORCE.ipynb 

R is implemented as a single value in the following code: 

```
discounts = [gamma**i for i in range(len(rewards)+1)]
R = sum([a*b for a,b in zip(discounts, rewards)])
```
However it should be implemented as discounted values as follow: 

```
        discounted_rewards = []
        for t in range(len(rewards)):
          Gt = 0 
          pwr = 0
          for r in rewards[t:]:
              Gt = Gt + gamma**pwr * r
              pwr = pwr + 1
          discounted_rewards.append(Gt)
        
        policy_loss = []
        for log_prob, Gt in zip(saved_log_probs, discounted_rewards):
          policy_loss.append(-log_prob * Gt)
```
This important correction is compatible with the REINFORCE algorithm and leads to a faster and more stable training as shown in the figure. 

![reinforce](https://user-images.githubusercontent.com/19908396/68528547-feea3f00-02fc-11ea-9d9a-24775024a385.PNG)

**Comment by lukysummer** [comment]
I agree. **Expected return G_t (sum of FUTURE awards)** should be multiplied with each log p(At|St), which decreases as t increases, not the cumulative reward/episode R which is same for all t.
